### PR TITLE
Add mypy support

### PR DESCRIPTION
**Pull request by kinnala**
Run mypy in CI.

### CODE REVIEW

File: skfem/element/discrete_field.py [hunk 1]
```diff
@@ -14,39 +14,21 @@ class DiscreteField(NamedTuple):
     hess: Optional[ndarray] = None
     hod: Optional[ndarray] = None
 
-    @property
```
Comment (kinnala) [inline skfem/element/discrete_field.py:17]: Backwards incompatible

File: skfem/assembly/form/linear_form.py [hunk 2]
```diff
@@ -16,30 +16,30 @@ class LinearForm(Form):
     """
 
     def assemble(self,
-                 u: Basis,
-                 v: Optional[Basis] = None,
+                 ubasis: Basis,
+                 vbasis: Optional[Basis] = None,
```
Comment (kinnala) [inline skfem/assembly/form/linear_form.py:20]: Backwards incompatible

**Comment by kinnala** [comment]
There are two "technically backwards-incompatible" changes highlighted above. @gdmcbain do you think such small changes warrant for a major version increment or not? `DiscreteField.f/df/ddf` should have been removed already before in `2.0.0` but seems I have forgotten them. There was also a typing-related bug I fixed in `Form` which revealed another issue in the argument naming. This forced me to rename `u` to `ubasis` because the keyword argument `u=...` was used in example 13, line 65. The change is basically renaming `v=...` keyword argument to `vbasis=...`. I have never used this keyword argument explicitly so that's why I think it's a minor change, not warranting for a major version increment.

**Comment by kinnala** [comment]
Decided to revert `DiscreteField.f/df/ddf` because that was easy enough.

**Review by gdmcbain — APPROVED** [review]
Great! I'm very pleased to see type-checking added to the CI.  

I note that no changes were required to `docs/examples`, so don't really think the changes to the API should effect any users; otherwise the examples should be expanded.  There are lots of changes in `tests` like `.f` &rarr; `.value` for a `DiscreteField`; I wonder why that's is exercised so much in the tests but not at all in the examples.

**Pull request merged**